pkg/segment/reader/record/recordreader.go

```diff
@@ -36,6 +36,171 @@ import (
+	var allCols map[string]bool
+	allCols, exists := writer.CheckAndGetColsForUnrotatedSegKey(segKey)
+	if !exists {
+		allCols, exists = segmetadata.CheckAndGetColsForSegKey(segKey, vTable)
+		if !exists {
+			return nil, toputils.TeeErrorf("getColsForSegKey: globalMetadata does not have segKey: %s", segKey)
+		}
+	}
+
+	// TODO: make the CheckAndGetColsForSegKey functions return a set instead
+	// of a map[string]bool so we don't have to do the conversion here
+	return toputils.MapToSet(allCols), nil
+}
+
+func ReadColForRRCs(segKey string, rrcs []*utils.RecordResultContainer, cname string, qid uint64) ([]utils.CValueEnclosure, error) {
+	switch cname {
+	case config.GetTimeStampKey():
+		return readTimestampForRRCs(rrcs)
+	case "_index":
+		return readIndexForRRCs(rrcs)
+	default:
+		return readUserDefinedColForRRCs(segKey, rrcs, cname, qid)
+	}
+}
+
+func readTimestampForRRCs(rrcs []*utils.RecordResultContainer) ([]utils.CValueEnclosure, error) {
+	result := make([]utils.CValueEnclosure, len(rrcs))
+	for i, rrc := range rrcs {
+		result[i] = utils.CValueEnclosure{
+			Dtype: utils.SS_DT_UNSIGNED_NUM,
+			CVal:  rrc.TimeStamp,
+		}
+	}
+
+	return result, nil
+}
+
+func readIndexForRRCs(rrcs []*utils.RecordResultContainer) ([]utils.CValueEnclosure, error) {
+	result := make([]utils.CValueEnclosure, len(rrcs))
+	for i, rrc := range rrcs {
+		result[i] = utils.CValueEnclosure{
+			Dtype: utils.SS_DT_STRING,
+			CVal:  rrc.VirtualTableName,
+		}
+	}
+
+	return result, nil
+}
+
+// All the RRCs must belong to the same segment.
+func readUserDefinedColForRRCs(segKey string, rrcs []*utils.RecordResultContainer,
+	cname string, qid uint64) ([]utils.CValueEnclosure, error) {
+
+	if len(rrcs) == 0 {
+		return nil, nil
+	}
+
+	err := fileutils.GLOBAL_FD_LIMITER.TryAcquireWithBackoff(1, 10, "readUserDefinedColForRRCs")
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs failed to get lock for opening 1 file; err=%v", qid, err)
+		return nil, err
+	}
+	defer fileutils.GLOBAL_FD_LIMITER.Release(1)
+
+	blockMetadata, err := writer.GetBlockSearchInfoForKey(segKey)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to get block metadata for segKey %s; err=%v", qid, segKey, err)
+		return nil, err
+	}
+
+	blockSummary, err := writer.GetBlockSummaryForKey(segKey)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to get block summary for segKey %s; err=%v", qid, segKey, err)
+		return nil, err
+	}
+
+	consistentCValLen := map[string]uint32{cname: utils.INCONSISTENT_CVAL_SIZE} // TODO: use correct value
+	sharedReader, err := segread.InitSharedMultiColumnReaders(segKey, map[string]bool{cname: true},
+		blockMetadata, blockSummary, 1, consistentCValLen, qid)
+	if err != nil {
+		log.Errorf("qid=%v, readUserDefinedColForRRCs: failed to initialize shared readers for segkey %v; err=%v", qid, segKey, err)
+		return nil, err
+	}
+	defer sharedReader.Close()
+	multiReader := sharedReader.MultiColReaders[0]
```

Comment:
why are we taking only the first reader (wouldn't this be just one csg file ?

Reply:
I don't fully understand the readers, so you may be right. I copied this part from `getRecordsFromSegmentHelper()` where we get the first; in the call to `InitSharedMultiColumnReaders()` we pass `1` as `numReaders`, so there should only be one reader returned, which is why I'm doing this. I thought that one reader would be able to read all the necessary columns, but I might be wrong

Reply:
ok, lets take a note of this, and once we are in the testing phase , then we can validate this